doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,240 @@
+
+✓ = 1st level API support, ● = supported by combining APIs, ✗ = no support
+
+1. Threads can wait until a software timer fires. Software timers accept NULL
+   as empty callback. Combined, this can be used to delay execution.
+2. Contiki provides with `ctimer_reset()`/`etimer_reset()` which restarts the
+   timer from the previous expiration time, allowing to easily achieve
+   periodic callbacks.
+3. The Rtimer library seems to run the callbacks in IRQ context and the Ctimer
+   library in thread context.
+4. The Etimer library posts events on timer expiration and a thread can block
+   until a specific event has been posted. Combined, this can be used to
+   delay execution
+5. Only available when using TOSTThreads
+6. `xtimer_periodic_wakeup()`/`ztimer_periodic_wakeup()` can be used, but
+   requires a dedicated thread allocated for the periodic task
+7. Mynwest uses a `struct` with a members `tv_sec` (64 bit) and `tv_usec`
+   (32 bit)
+8. `ztimer_now()` returns the current time of the given clock. This can be used
+   as system timer
+9. "Portable System Time" means that two nodes can exchange time stamps over
+   network and they still refer to the same point in time, assuming the clocks
+   have been synchronized.
+
+# 2. Requirements
+
+The following lists of requirements consist of qualitative and quantitative
+requirements. For the quantitative requirements, any improvement will enable
+more scenarios to use the timer API. Thus, no meaningful lower bounds can be
+specified. Instead, a set of benchmarks will be developed that allow to
+continuously improve the performance and to detect any possible regressions of
+future changes.
+
+## Feature Requirements
+
+- Software Timers
+    - Upper bound for number of timers only given by available resources
+    - APIs to set and abort timers
+    - Support for periodic timers (support via explicit calls to some
+      `timer_reset()` like in Contiki is fine)
+    - Long term low resolution software timers that play well with power management
+    - Short term high resolution software timers
+- Delay of program execution
+    - API to block calling thread for long periods and allows other threads to run
+    - Long term delays must play well with power management
+    - Highly precise short term delays (e.g. for bit banging)
+- System time
+    - APIs to get the current system time in reasonable, hardware independent
+      units
+        - At least: seconds, milli seconds
+    - UNIX epoch as common reference point for high interoperability
+    - Support for 64 bit time stamps
+        - Devices running for detects might want to log events with milli second
+          resolution timestamps on flash / sd card / ...
+        - Note: The rest of the API can be 32 bit, only for time stamps 64 bit
+          support is mandatory
+    - Optional feature to auto-initialize system time from RTC/RTT during boot
+    - APIs to allow implementing network time synchronization
```

Comment:
+1 here

With LoRaWAN there are some commands to get the absolute timestamp from the network server (with < 10ms of accuracy). Implementing such an API for LoRaWAN would be straightforward.